osrf/mbzirc
# Add boundary penalty to Game Logic

Extend GameLogic plugin to support paring a geofence boundary for the competition area:

For example, this sets a geofence boundary of ~10km2 in the Coast environment
```
      <geofence>
        <center>0 0 10.96</center>
        <size>3162.28 3162.28 221.92</size>
      </geofence>
```
This area is highlighted in green in the image shown in https://github.com/osrf/mbzirc/pull/38

When a robot moves beyond the competition boundary, a time penalty is applied. There is a buffer distance for it to move back by before another penalty is incurred. This is so that we don't immediately trigger another penalty when the USV oscillates due to waves.

Other changes:
* other time penalty types are included in the enum but are not used yet.
* boundary penalty events are logged
* added test
* added IMU to USV - so that it can be recognized as a robot by game logic plugin (any entity with a sensor).

To test:

1. launch coast world
    ```
    ros2 launch ros_ign_gazebo ign_gazebo.launch.py ign_args:="-v 4 -r coast.sdf"
    ```

1. Spawn the USV

    ```
    ros2 launch mbzirc_ign spawn.launch.py name:=usv world:=coast model:=usv type:=usv x:=-1450 y:=0 z:=0.3 R:=0 P:=0 Y:=0
    ```

1. In a separate terminal, monitor the `events.yml` log file, e.g.

    ```
    tail -f /tmp/ign/mbzirc/logs/events.yml
    ```

1. Move the USV out of bounds and see 2 events triggered: `started` and `exceed_boundary_1` in `events.yml`. A time penalty of 300 is also added to the `total_score` (total score is elapsed sim time + penalty).

    ```
    ign service -s /world/coast/set_pose --reqtype ignition.msgs.Pose --reptype ignition.msgs.Boolean --timeout 300 --req 'name: "usv", position: {y: 2000.0}'
    ```

1. Move the USV back within bounds, e.g. 

    ```
    ign service -s /world/coast/set_pose --reqtype ignition.msgs.Pose --reptype ignition.msgs.Boolean --timeout 300 --req 'name: "usv", position: {y: 1000.0}'
    ```


1. Move the USV out of bounds again and see 2 more events triggered: `exceed_boundary_2` and `finished`. The 2nd `exceed_boundary` causes termination of the run.

    ```
    ign service -s /world/coast/set_pose --reqtype ignition.msgs.Pose --reptype ignition.msgs.Boolean --timeout 300 --req 'name: "usv", position: {y: 2000.0}'
    ```
